classes/manager.php

```diff
@@ -144,5 +144,81 @@ public static function is_factorid_valid($factorname, $factorid, $user) {
 
         return false;
     }
+
+    /**
+     * Function to display to the user that they cannot login, then log them out.
+     *
+     * @return void
+     */
+    public static function cannot_login() {
+        self::mfa_logout();
+        print_error('error:notenoughfactors', 'tool_mfa', new \moodle_url('/'));
+    }
+
+    /**
+     * Logout user.
+     *
+     * @return void
+     */
+    public static function mfa_logout() {
+        $authsequence = get_enabled_auth_plugins();
+        foreach ($authsequence as $authname) {
+            $authplugin = get_auth_plugin($authname);
+            $authplugin->logoutpage_hook();
+        }
+        require_logout();
+    }
+
+    /**
+     * Function to check the overall status of a user's authentication.
+     *
+     * @return mixed a STATE variable from plugininfo
+     */
+    public static function check_status() {
+        global $SESSION;
+
+        // Check for any instant fail states.
+        $factors = \tool_mfa\plugininfo\factor::get_active_user_factor_types();
+        $fail = false;
+        foreach ($factors as $factor) {
+            if ($factor->get_state() == \tool_mfa\plugininfo\factor::STATE_FAIL) {
+                $fail = true;
+            }
+        }
+        if ($fail) {
+            return \tool_mfa\plugininfo\factor::STATE_FAIL;
+        }
+
+        // Check for passing state. If found, ensure that session var is set.
+        if (isset($SESSION->tool_mfa_authenticated) && $SESSION->tool_mfa_authenticated) {
+            return \tool_mfa\plugininfo\factor::STATE_PASS;
+        } else if (self::passed_enough_factors()) {
+            self::set_pass_state();
+            return \tool_mfa\plugininfo\factor::STATE_PASS;
+        }
+
+        // Else return neutral state.
+        return \tool_mfa\plugininfo\factor::STATE_NEUTRAL;
+    }
+
+    /**
+     * Checks whether user has passed enough factors to be allowed in
+     */
+    public static function passed_enough_factors() {
+        $totalweight = self::get_total_weight();
+        if ($totalweight >= 100) {
+            return true;
+        }
+
+        return false;
+    }
+
+    public static function set_pass_state() {
+        global $SESSION, $USER;
+        $SESSION->tool_mfa_authenticated = true;
```

Comment:
Just wondering if we should test if we are already passed and not emit the event the second time?